### PR TITLE
refactor(connector): replace all `RwError` usage with `anyhow` temporarily

### PR DESCRIPTION
**Pull request by BugenZhao**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

Replace all `RwError` usages in the `connector` crate to `anyhow` utilities as a temporary solution to eliminate the usage of `RwError`. After this PR, we're able to move `RwError` into the `frontend` crate!

The refactor is always making things better as:

- The only variants of `RwError` we were actually using were `ProtocolError` and `InternalError`. In fact, they was chosen not because they were the best fit, but because they accepted a `String` as a parameter so developers can easily construct them and make it compile. 😄 As a result, we can replace them with `anyhow::Error` without losing any semantics.

- `RwError` always captures the backtrace, which may lead to unexpected performance issues sometimes. So does the `anyhow::Error` in the nightly channel, so it doesn't make things worse. For most of the critical paths, we've already used other error types (like `AccessError` in parsing) to avoid the backtrace. (#14874)

- `anyhow::Error` are already massively used in the `connector` crate.

We plan to do further refactor in the future based on this PR:

- Wrap `anyhow::Error` into a new type to include the type information, so that the boundary between the `connector` crate and dependent crates can be much clearer.

- Find a way to avoid backtrace capturing in the `anyhow::Error`.

## Checklist

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

## Documentation

- [ ] My PR needs documentation updates. (Please use the **Release note** section below to summarize the impact on users)

## Release note

If this PR includes changes that directly affect users or other significant modifications relevant to the community, kindly draft a release note to provide a concise summary of these changes. Please prioritize highlighting the impact these changes will have on users.

</details>

### CODE REVIEW

**Comment by BugenZhao** [comment]
Current dependencies on/for this PR:
* `main`
  * **PR #14950** <a href="https://app.graphite.dev/github/pr/risingwavelabs/risingwave/14950?utm_source=stack-comment-icon" target="_blank">Graphite</a>  👈

This [stack of pull requests](https://stacking.dev/?utm_source=stack-comment) is managed by [Graphite](https://app.graphite.dev/github/pr/risingwavelabs/risingwave/14950?utm_source=stack-comment).

**Review by xxchan — COMMENTED** [review]
LSTM

**Review by xxchan — APPROVED** [review]

File: src/connector/src/parser/json_parser.rs [hunk 1]
```diff
@@ -134,11 +131,12 @@ impl JsonParser {
...
             Ok(())
         } else {
-            Err(RwError::from(ErrorCode::InternalError(format!(
+            // TODO(error-handling): multiple errors
+            bail!(
                 "failed to parse {} row(s) in a single json message: {}",
                 errors.len(),
-                errors.iter().join(", ")
-            ))))
+                errors.iter().format(", ")
```
Comment (stdrc) [inline src/connector/src/parser/json_parser.rs:138]: What's the difference between `join` and `format`? Is the latter a better practice? `join` looks more intuitive to me actually.

File: src/connector/src/parser/mod.rs [hunk 2]
```diff
@@ -729,7 +730,8 @@ async fn into_chunk_stream<P: ByteStreamSourceParser>(mut parser: P, data_stream
 }
 
 pub trait AccessBuilder {
-    async fn generate_accessor(&mut self, payload: Vec<u8>) -> Result<AccessImpl<'_, '_>>;
+    // TODO(eh)
```
Comment (stdrc) [inline src/connector/src/parser/mod.rs:733]:
```suggestion
    // TODO(error-handling)
```

Comment (stdrc) [reply src/connector/src/parser/mod.rs:733]: Maybe no one will known what `eh` is later. Also for several other places.

File: src/connector/src/parser/util.rs [hunk 3]
```diff
@@ -69,9 +67,7 @@ macro_rules! only_parse_payload {
         if let Some(payload) = $payload {
             $self.parse_inner(payload, $writer).await
         } else {
-            Err(RwError::from(ErrorCode::InternalError(
-                "Empty payload with nonempty key".into(),
-            )))
+            risingwave_common::bail!("empty payload with non-empty key")
```
Comment (stdrc) [inline src/connector/src/parser/util.rs:70]:
```suggestion
            bail!("empty payload with non-empty key")
```

**Review by stdrc — APPROVED** [review]
LSTM!

File: src/connector/src/parser/json_parser.rs (hunk 1 above)
Comment (xxchan) [reply src/connector/src/parser/json_parser.rs:138]: `format` saves an unnecessary string allocation.

File: src/connector/src/parser/mod.rs (hunk 2 above)
Comment (BugenZhao) [reply src/connector/src/parser/mod.rs:733]: Oops, I must have forgotten to clean them up.

File: src/connector/src/parser/util.rs (hunk 3 above)
Comment (BugenZhao) [reply src/connector/src/parser/util.rs:70]: Since it'll be expanded, I prefer qualified path here to make it self-contained.

Comment (stdrc) [reply src/connector/src/parser/util.rs:70]: Oh, didn't realize it's in a macro.

**Pull request merged**